crates/bevy_animation/src/graph.rs

```diff
@@ -0,0 +1,399 @@
+use bevy_utils::BoxedFuture;
+use petgraph::graph::{DiGraph, NodeIndex};
+use ron::de::SpannedError;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use crate::AnimationClip;
+
+/// A graph structure that describes how animation clips are to be blended
+/// together.
+///
+/// Applications frequently want to be able to play multiple animations at once
+/// and to fine-tune the influence that animations have on a skinned mesh. Bevy
+/// uses an *animation graph* to store this information. Animation graphs are a
+/// directed acyclic graph (DAG) that describes how animations are to be
+/// weighted and combined together. Every frame, Bevy evaluates the graph from
+/// the root and blends the animations together in a bottom-up fashion to
+/// produce the final pose.
+///
+/// There are two types of nodes: *blend nodes* and *clip nodes*, both of which
+/// can have an associated weight. Blend nodes have no associated animation clip
+/// and simply affect the weights of all their descendant nodes. Clip nodes
+/// specify an animation clip to play. When a graph is created, it starts with
+/// only a single blend node, the root node.
+///
+/// For example, consider the following graph:
+///
+/// ```text
+/// ┌────────────┐                                      
+/// │            │                                      
+/// │    Idle    ├─────────────────────┐                
+/// │            │                     │                
+/// └────────────┘                     │                
+///                                    │                
+/// ┌────────────┐                     │  ┌────────────┐
+/// │            │                     │  │            │
+/// │    Run     ├──┐                  ├──┤    Root    │
+/// │            │  │  ┌────────────┐  │  │            │
+/// └────────────┘  │  │   Blend    │  │  └────────────┘
+///                 ├──┤            ├──┘                
+/// ┌────────────┐  │  │    0.5     │                   
+/// │            │  │  └────────────┘                   
+/// │    Walk    ├──┘                                   
+/// │            │                                      
+/// └────────────┘                                      
+/// ```
+///
+/// In this case, assuming that Idle, Run, and Walk are all playing with weight
+/// 1.0, the Run and Walk animations will be equally blended together, then
+/// their weights will be halved and finally blended with the Idle animation.
+/// Thus the weight of Run and Walk are effectively half of the weight of Idle.
+///
+/// Animation graphs are assets and can be serialized to and loaded from [RON]
+/// files. Canonically, such files have an `.animgraph.ron` extension.
+///
+/// The animation graph implements [RFC 51]. See that document for more
+/// information.
+///
+/// [RON]: https://github.com/ron-rs/ron
+///
+/// [RFC 51]: https://github.com/bevyengine/rfcs/blob/main/rfcs/51-animation-composition.md
+#[derive(Asset, Reflect, Clone, Debug, Serialize)]
```

Comment:
This should also have the corresponding `reflect` attribute for Serialize, Deserialize, and Debug.